### PR TITLE
utilities: add a collection of associated testing tools

### PR DESCRIPTION
**Pull request by andrewbonney**
We've been working on a few quick tools which are initially intended to assist with the JT-NM event in August, but may be more generally useful. I haven't advertised these from the main README yet as they're likely to undergo a little more tweaking, but this is the initial contribution.

### CODE REVIEW

**Comment by garethsb-sony** [comment]
We've been experimenting with the gsheets-import utility integrated into a CI pipeline. It's working pretty well. We use the filename to indicate the git commit hash and build config.

A couple of issues:
* Needs something in Row 1 or it crashes out (i.e. doesn't work with a completely empty sheet!)
* Doesn't seem to handle a new test case being introduced, with the same result?

In both cases, I think it would be nice to automatically insert a row of column headings. That means when the tests change you'd get a new heading row halfway down the sheet, but I think that's the best we can do in these circumstances.

I set up 'conditional formatting' to colour the test case cells red/yellow/green/blue as appropriate, not sure if it would make sense to do this in code. It would be nice to add some row totals (total number of tests, number of red/yellow/green/blue), haven't yet tried to see if I can do this and have the formulas be applied to newly inserted rows, or whether this needs script change.

To me it seems as if making the Google Sheet URL a command-line parameter would be appropriate so that the script can work out of the box. (I'd probably do the same for the credentials json, rather than relying on a filename.) By all means let's keep the internal variables/hard-coded filename as a fallback so that you can edit those like the test suite Config.py if preferred.

File: utilities/dut-data-exporter/dutDataExporter.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+def url_port_number(url):
+    result = urlparse(url)
+    try:
+        if result.port:
+            return result.port
+        else:
+            return 80
+    except ValueError:
+        print("IPv6 address could not be parsed: {}".format(url))
```
Comment (garethsb-sony) [inline utilities/dut-data-exporter/dutDataExporter.py]: Why does this refer to IPv6?

File: utilities/dut-data-exporter/dutDataExporter.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+    is04Port = args.port
+    is05Port = None
+    is08Port = None
+    node_id = None
+    senders = []
+    receivers = []
+    mac_addresses = []
+
+    base_url = "http://{}:{}/x-nmos/node/v1.2/".format(args.ip, args.port)
```
Comment (garethsb-sony) [inline utilities/dut-data-exporter/dutDataExporter.py]: This was fine for JT-NM Tested August 2019, but should version now be made configurable on the command line, like `--version v1.3`?

File: utilities/dut-data-exporter/dutDataExporter.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    node_id = response.json()['id']
+    print("Host: {}".format(response.json()['description']))
+
+    # Devices
+    url = base_url + "devices/"
+    response = requests.get(url)
+    json_data = response.json()
+    for d in json_data[0]["controls"]:
+        if d['type'] == "urn:x-nmos:control:sr-ctrl/v1.0":
+            is05Port = url_port_number(d['href'])
```
Comment (garethsb-sony) [inline utilities/dut-data-exporter/dutDataExporter.py]: Here too, IS-05 v1.1 should be configurable?

File: utilities/dut-data-exporter/dutDataExporter.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    mac_addresses = []
+
+    base_url = "http://{}:{}/x-nmos/node/v1.2/".format(args.ip, args.port)
+
+    # Self
+    url = base_url + "self/"
+    response = requests.get(url)
+    mac_addresses = response.json()['interfaces']
+    for address in mac_addresses:
+        if type(address['port_id']) == str:
```
Comment (garethsb-sony) [inline utilities/dut-data-exporter/dutDataExporter.py]: Mix of single quotes and double quotes in this file is a little off-putting to my punctilious left brain. :-)

File: utilities/gsheets-import/bulk-import.sh [hunk 5]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+for filename in ~/Downloads/nmos-test-results/*.json; do
+    printf "$filename \n"
```
Comment (garethsb-sony) [inline utilities/gsheets-import/bulk-import.sh]: What's the space for?

File: utilities/gsheets-import/resultsImporter.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
+#!/usr/bin/python
```
Comment (garethsb-sony) [inline utilities/gsheets-import/resultsImporter.py]: I'd love this tool to be just a little more flexible. But that's covered by #370.

File: utilities/is-05-control/README.md [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+```
+python3 is05Control.py --receiver --ip <hostname or IP> --port 80 --uuid <Receiver ID> --request example_receiver_patch.json --sdp sdp.sdp
+```
+
+Example call to enable sender
+```
+python3 is05Control.py --sender --ip <hostname or IP> --port 80 --uuid <Sender ID> --request example_sender_patch.json
+```
+
+NOTE: R&D proxy cannot handle PATCH request, so make sure no_proxy value is set correctly
```
Comment (garethsb-sony) [inline utilities/is-05-control/README.md]: Explain in general terms or remove this note?

File: utilities/is-05-control/is05Control.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+parser.add_argument("-d", "--disable", action="store_true", help="Set master_enable=False")
+parser.add_argument("-r", "--request", required=True, help="Patch file to be sent in the request")
+parser.add_argument("-s", "--sdp", help="SDP file to be sent in the request")
+parser.add_argument("-u", "--uuid", required=True, help="UUID of resource to be configured")
+args = parser.parse_args()
+
+# Configure for Sender or Receiver
+if args.sender:
+    print("Configuring NMOS Sender using IS-05")
+    url = "http://{}:{}/x-nmos/connection/v1.0/single/senders/{}/staged".format(args.ip, args.port, args.uuid)
```
Comment (garethsb-sony) [inline utilities/is-05-control/is05Control.py]: Another place where I think we now ought to allow ``--version v1.1`` or something on the command line to configure this.

File: utilities/uuid-checker/README.md [hunk 9]
```diff
@@ -0,0 +1,20 @@
+# NMOS UUID Checker
+Command line tool to check that NMOS IS-04 v1.2 UUIDs persist over a reboot
```
Comment (garethsb-sony) [inline utilities/uuid-checker/README.md]: Again, could we now allow configuration of spec version?

File: utilities/dut-data-exporter/dutDataExporter.py (hunk 1 above)
Comment (andrewbonney) [reply utilities/dut-data-exporter/dutDataExporter.py]: I assume urlparse has some sort of issue with IPv6, but I'll have to check with James when he's back in.

Comment (garethsb-sony) [reply utilities/dut-data-exporter/dutDataExporter.py]: OK, but presumably a `ValueError` is also raised for other unparseable hosts?

Comment (andrewbonney) [reply utilities/dut-data-exporter/dutDataExporter.py]: Ah I see, yes I can adjust the message for that purpose.

**Pull request merged**